src/Microsoft.AspNet.OData.Shared/Builder/EdmModelHelperMethods.cs

```diff
@@ -638,13 +652,64 @@ private static void AddCoreVocabularyAnnotations(this EdmModel model, IEnumerabl
             if (edmProperties.Any())
             {
                 IEdmCollectionExpression collectionExpression = new EdmCollectionExpression(edmProperties.Select(p => new EdmPropertyPathExpression(p.Name)).ToArray());
-                IEdmTerm term = Microsoft.OData.Edm.Vocabularies.V1.CoreVocabularyModel.ConcurrencyTerm;
+                IEdmTerm term = CoreVocabularyModel.ConcurrencyTerm;
                 EdmVocabularyAnnotation annotation = new EdmVocabularyAnnotation(target, term, collectionExpression);
                 annotation.SetSerializationLocation(model, EdmVocabularyAnnotationSerializationLocation.Inline);
                 model.SetVocabularyAnnotation(annotation);
             }
         }
 
+        private static void AddDerivedTypeConstraintAnnotationForProperties(this EdmModel model, EdmTypeMap edmTypeMap)
+        {
+            Dictionary<IEdmProperty, PropertyConfiguration>
+                propertyConfiguration = edmTypeMap.EdmPropertyConfigurations;
+
+            foreach (var pair in propertyConfiguration)
+            {
+                if (pair.Value.DerivedTypeConstraints.Any())
+                {
+                    model.AddDerivedTypeConstraintAnnotation(pair.Key, edmTypeMap.EdmTypes, pair.Value.DerivedTypeConstraints.ToList());
+                }
+            }         
+        }
+
+        private static void AddDerivedTypeConstraintAnnotationForNavigationSource(this EdmModel model, IEdmVocabularyAnnotatable target,
+            NavigationSourceConfiguration navigationSourceConfiguration, EdmTypeMap edmTypeMap)
+        {
+            if (navigationSourceConfiguration != null)
+            {
+                model.AddDerivedTypeConstraintAnnotation(target,
+                    edmTypeMap.EdmTypes, navigationSourceConfiguration.DerivedTypeConstraints.ToList());
+            }
+        }
+
+        internal static void AddDerivedTypeConstraintAnnotation(this EdmModel model, IEdmVocabularyAnnotatable target, 
+            Dictionary<Type, IEdmType> edmTypeMap, List<Type> types)
+        {
+            IEdmTerm term = ValidationVocabularyModel.DerivedTypeConstraintTerm;
+            if (types.Count == 0)
+            {
+                return;
+            }
+
+            List<EdmStringConstant> collectionConstants = new List<EdmStringConstant>();
+            foreach (var type in types)
+            {
+                if (edmTypeMap.ContainsKey(type))
+                {
+                    collectionConstants.Add(new EdmStringConstant(edmTypeMap[type].FullTypeName()));
+                }
+                else
+                {
+                    throw Error.InvalidOperation("Cannot find EdmType corresponding to {0}", type);
+                }
+            }
+
+            var collectionExpression = new EdmCollectionExpression(collectionConstants);
+            EdmVocabularyAnnotation annotation = new EdmVocabularyAnnotation(target, term, collectionExpression);
```

Comment:
>EdmVocabularyAnnotation annotation = new EdmVocabularyAnnotation(target, term, collectionExpression); [](start = 12, length = 101)

Discussion: Maybe need configuration to set the annotation location (Inline or outofLine)